### PR TITLE
Backend user routes

### LINKED ISSUE #90: User Registration functionality
## Feature Description
Code for user registration, which involves creating a JWT token on sign up and returning it with the new user object to the front end

blocked by: #82

## Problem You're Trying to Solve
Users must be able to register for the website in order to use it

## Related External Resources
[section 4.4.1](https://docs.google.com/document/d/1DNqexMBf4MpYPCxWsT3plUDAjDSbdJwUzw-UyZX-vDs/edit#heading=h.vbv6udt3y8om)

### PR DESCRIPTION
**Pull request by Scott-Kenning**
# Description
All the backend user routes, including login/logout

Closes #90 #91

## How to Test
Run it

## Checklist
- [x] The code includes tests if relevant
- [x] I have *actually* self-reviewed my changes and done QA

### CODE REVIEW

**Review by TobySMurray — APPROVED** [review]
Lovely code, lgtm!

File: apps/backend/user/src/login.ts
Comment (kylestang) [inline apps/backend/user/src/login.ts:53]:
```suggestion
    return res.status(500).json({ error: "Something went wrong" });
```

File: apps/backend/user/src/patchUser.ts
Comment (kylestang) [inline apps/backend/user/src/patchUser.ts:22]: Shouldn't matter but js type conversion scares me
```suggestion
  if (id === 1) {
```

File: apps/backend/user/src/login.ts
Comment (abookout) [inline apps/backend/user/src/login.ts:20]: resops (Max) might not like this because it allows a malicious agent to sniff out which emails have an account on the site, making it easier to target accounts for cracking their passwords etc

File: apps/backend/user/src/logout.ts
Comment (abookout) [inline apps/backend/user/src/logout.ts:4]: Is this being left for resops to finish?

File: apps/backend/user/src/patchUser.ts
Comment (kylestang) [inline apps/backend/user/src/patchUser.ts:102]:
```suggestion
    return res.status(500).json({ error: "Something went wrong" });
```

File: apps/backend/user/src/patchUser.ts
Comment (abookout) [inline apps/backend/user/src/patchUser.ts:61]: hashedPassword is not used in the patch

**Review by kylestang — CHANGES_REQUESTED** [review]
Mostly good, a few changes to go through.

Nit: Sometimes the endpoints use `console.log(err)` and other times `console.error(err)`. Would be nice if those were consistent

File: apps/backend/user/src/login.ts
Comment (Scott-Kenning) [reply apps/backend/user/src/login.ts:20]: I'll change both catches to just return "invalid email or password"

File: apps/backend/user/src/logout.ts
Comment (Scott-Kenning) [reply apps/backend/user/src/logout.ts:4]: correct

File: apps/backend/user/src/patchUser.ts
Comment (Scott-Kenning) [reply apps/backend/user/src/patchUser.ts:61]: good catch

File: apps/backend/user/cypress/fixtures/example.json
Comment (Scott-Kenning) [reply apps/backend/user/cypress/fixtures/example.json:1]: @islandj123

File: apps/backend/user/src/patchUser.ts
Comment (kylestang) [reply apps/backend/user/src/patchUser.ts:102]: Also, the error isn't logged on this endpoint like it is on the other

File: apps/backend/user/cypress/fixtures/example.json
Comment (Scott-Kenning) [reply apps/backend/user/cypress/fixtures/example.json:1]: I removed this in the other PR

File: apps/backend/user/cypress.config.ts
Comment (MNThomson) [inline apps/backend/user/cypress.config.ts:5]: This being hard-coded makes me unhappy, but not sure there's a better solution

File: apps/backend/user/cypress.config.ts
Comment (MNThomson) [inline apps/backend/user/cypress.config.ts:6]: These aren't e2e tests being written

File: apps/backend/user/cypress/fixtures/example.json
Comment (MNThomson) [inline apps/backend/user/cypress/fixtures/example.json:5]: This file?

File: apps/backend/user/src/createUser.ts
Comment (MNThomson) [inline apps/backend/user/src/createUser.ts:12]: I didn't realize all these fields were on the Create user endpoint

File: apps/backend/user/src/createUser.ts
Comment (MNThomson) [inline apps/backend/user/src/createUser.ts:42]: Errrrrrm, hashed password please
```suggestion
    INSERT INTO users (username, email, hashedPassword, name, bio, profile_pic_url, verified)
```

File: apps/backend/user/src/createUser.ts
Comment (MNThomson) [inline apps/backend/user/src/createUser.ts:54]: These values just don't match the sql return statement at all

File: apps/backend/user/src/deleteUser.ts
Comment (MNThomson) [inline apps/backend/user/src/deleteUser.ts:14]: Why can't we delete the "base user"? Also, what is the "base user"?

File: apps/backend/user/src/deleteUser.ts
Comment (MNThomson) [inline apps/backend/user/src/deleteUser.ts:39]: Is user ID 1 the "deleted user" user? I believe we have in the db a cascade delete so that when a user is deleted, their postings and message conversations are deleted (think GDPR, right to be forgotten, etc)

File: apps/backend/user/src/deleteUser.ts
Comment (MNThomson) [inline apps/backend/user/src/deleteUser.ts:54]: Error handling is good but can we get to this case? They have to have a valid JWT but maybe it hasn't expired yet? This seems like a case that the general tey except could handle

File: apps/backend/user/src/getUser.ts
Comment (MNThomson) [inline apps/backend/user/src/getUser.ts:11]: Let's request specific properties, using `*` is an antipattern

File: apps/backend/user/src/login.ts
Comment (MNThomson) [inline apps/backend/user/src/login.ts:21]: This exposes us to a timing attack. An attacker can try email/password combos and can check whether or not an email exists in the db by how long the response takes (short if not signed up since it returns early, long if the email exists since then more cryptographic operations are performed on the password).

The easiest way to mitigate this is to always make a bcrypt comparison either against the hashed password in the db or a static string of a random value.

File: apps/backend/user/src/login.ts
Comment (MNThomson) [inline apps/backend/user/src/login.ts:33]:
```suggestion
    if (!isPasswordValid || !iuserExists) {
      return res.status(401).json({ error: "Invalid email or password" });
    }
```

File: apps/backend/user/src/login.ts
Comment (MNThomson) [inline apps/backend/user/src/login.ts:41]: I will be implementing the auth cookie Tue night, let's remove this (since there's a lot of work needed in the JWT area)

File: apps/backend/user/src/models/user.ts
Comment (MNThomson) [inline apps/backend/user/src/models/user.ts:12]: Not sure if this is used in multiple places, but all the user endpoints have a slightly different return type

File: apps/backend/user/src/patchUser.ts
Comment (MNThomson) [inline apps/backend/user/src/patchUser.ts:15]: We will need password checks (8 chars, lower/upper, etc) here also

Comment (MNThomson) [inline apps/backend/user/src/patchUser.ts:15]: Also, no way they are allowed to update their own verified status
```suggestion
  const { username, password, name, bio, profilePictureUrl } =
```

File: apps/backend/user/src/patchUser.ts
Comment (MNThomson) [inline apps/backend/user/src/patchUser.ts:33]:
```suggestion
```

File: apps/backend/user/src/patchUser.ts
Comment (MNThomson) [inline apps/backend/user/src/patchUser.ts:76]:
```suggestion
```

File: apps/backend/user/src/patchUser.ts
Comment (MNThomson) [inline apps/backend/user/src/patchUser.ts:68]: This return seems to have stuff that's never used

File: apps/data/database/populatedb.sql
Comment (MNThomson) [inline apps/data/database/populatedb.sql:2]: Ah, understood. Still confused why we aren't just allowing the cascade delete to do its thing though. Do we really want that stuff staying around?

**Review by MNThomson — CHANGES_REQUESTED** [review]

File: apps/data/database/populatedb.sql
Comment (Scott-Kenning) [reply apps/data/database/populatedb.sql:2]: Yes, we don't want messages or reviews (primarily messages) to be removed when a user is deleted

File: apps/backend/user/src/createUser.ts
Comment (Scott-Kenning) [reply apps/backend/user/src/createUser.ts:42]: the hashed password is passed in the value, to the field named password

Comment (MNThomson) [reply apps/backend/user/src/createUser.ts]: Am idot

File: apps/backend/user/src/createUser.ts
Comment (MNThomson) [inline apps/backend/user/src/createUser.ts:14]: No validation on name?

File: apps/backend/user/src/createUser.ts
Comment (MNThomson) [inline apps/backend/user/src/createUser.ts:47]: And reason for `false` to not be in the sql query?

```suggestion
  const values = [username, email, hashedPassword, name];
```

File: apps/backend/user/src/createUser.ts
Comment (MNThomson) [inline apps/backend/user/src/createUser.ts:60]: Nothing required, but why are we combing `await` and `.then`?

File: apps/backend/user/src/deleteUser.ts
Comment (MNThomson) [inline apps/backend/user/src/deleteUser.ts:46]: Future work to combine into one query. If one fails, the entire operation should rollback

File: apps/backend/user/src/login.ts
Comment (kylestang) [reply apps/backend/user/src/login.ts]: I considered this and the old issue where the error messages distinguished between wrong username and wrong password. I don't think it matters if people know a username/email is in the DB

File: apps/backend/user/src/login.ts
Comment (MNThomson) [inline apps/backend/user/src/login.ts:26]: This won't run if the user doesn't exist. Still vulnerable to a timing attack.

THIS IS PROBABLY NOT VALID CODE but serves as a reference
```suggestion
      let maybePassword = user.password || "";
      isPasswordValid = await bcrypt.compare(password, maybePassword);
```

**Review by MNThomson — APPROVED** [review]
I'm proactively approving this to unblock you if I don't get back to this tonight. Merge it when you feel it's ready

File: apps/backend/user/src/deleteUser.ts
Comment (Scott-Kenning) [reply apps/backend/user/src/deleteUser.ts:36]: this is the buyer ID. for deleted users, the delete will cascade and delete the listings

**Comment by MNThomson** [comment]
LGTM on resolving timing attack

**Review by kylestang — APPROVED** [review]
Approving to get this moved along, but I think we should review the user deletion process in sprint 3. There are a bunch of edge cases I'm not sure we have handled

**Review by abookout — APPROVED** [review]

**Pull request merged**
